router/diag_cntl.c

```diff
@@ -556,6 +598,118 @@ void diag_cntl_set_buffering_mode(struct peripheral *perif, int mode)
 	}
 }
 
+static int diag_query_pd_name(char *process_name, char *search_str)
+{
+	if (!process_name)
+		return -EINVAL;
+
+	if (strstr(process_name, search_str))
+		return 1;
+
+	return 0;
+}
+
+int diag_add_diag_id_to_list(uint8_t diag_id, char *process_name, struct peripheral *perif)
+{
+	struct diag_id_tbl_t *new_diagid = NULL;
+	int process_len = 0;
+	int ret;
+
+	if (!process_name || diag_id == 0)
+		return -EINVAL;
+
+	new_diagid = malloc(sizeof(struct diag_id_tbl_t));
+	if (!new_diagid)
+		return -ENOMEM;
+
+	process_len = strlen(process_name);
+	new_diagid->process_name = malloc(process_len + 1);
+	if (!new_diagid->process_name) {
+		free(new_diagid);
+		new_diagid = NULL;
+		return -ENOMEM;
+	}
+
+	new_diagid->diag_id = diag_id;
+	(void)strlcpy(new_diagid->process_name, process_name, process_len + 1);
+	list_add(&g_diag_id_list, &new_diagid->link);
+
+	return 0;
+}
+
+int diag_get_diag_id(char *name, uint8_t *diag_id)
+{
+	struct diag_id_tbl_t *diagid = NULL;
+	struct list_head *item;
+
+	if (!name || !diag_id)
+		return -EINVAL;
+
+	list_for_each(item, &g_diag_id_list) {
+		diagid = list_entry(item, struct diag_id_tbl_t, link);
+		if (!strcmp(diagid->process_name, name)) {
+			*diag_id = diagid->diag_id;
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+static int diag_cntl_process_diag_id(struct peripheral *peripheral, struct diag_cntl_hdr *hdr, size_t len)
+{
+	struct diag_cntl_cmd_diag_id *pkt = to_cmd_diagid(hdr);
+	struct diag_cntl_cmd_diag_id rsp;
+	struct diag_cntl_cmd_diag_id_v2 *pkt_v2 = to_cmd_diag_id_v2(hdr);
+	uint32_t version = 0, feature_len = 0;
+	char *process_name = NULL;
+	uint8_t local_diag_id = 0;
+	static uint8_t diag_id = DIAG_ID_APPS;
+	int ret;
+
+	if (pkt == NULL || len == 0)
+		return -EINVAL;
+
+	version = pkt->version;
+	if (peripheral->diagid_v2_feature_flag && version > DIAGID_VERSION_1) {
+		if (len < (sizeof(struct diag_cntl_cmd_diag_id_v2 ) -
+			(MAX_DIAGID_STR_LEN - MIN_DIAGID_STR_LEN))) {
+			return -EINVAL;
+		}
+		feature_len = (uint32_t)pkt_v2->feature_len;
+		process_name = (char *)&pkt_v2->feature_len +
+			sizeof(feature_len) + feature_len;
+	} else {
+		if (len < (sizeof(struct diag_cntl_cmd_diag_id) -
+			(MAX_DIAGID_STR_LEN - MIN_DIAGID_STR_LEN))) {
+			return -EINVAL;
+		}
+		process_name = (char*)&pkt->process_name;
+	}
+
+	ret = diag_get_diag_id(process_name, &local_diag_id);
+	if (!ret) {
+		if (version == DIAGID_VERSION_3)
+			diag_id = pkt_v2->diag_id;
```

Comment:
Since diag_id is static and being treated like a global counter. We probably dont want set diag_id to pkt_v2->diag_id.
I think we want to set local_diag_id here and leave diag_id alone.

Reply:
Agreed, downstream is likely incorrect. We should find a saner solution for upstream. It is likely not causing any issues because all the DSPs were upgraded to v3 at the same time

Reply:
If v3 packets with diagid 0 is received, then this is a peripheral bootup and values assigned same as non-v3.
But if there is a value in v3, then its diag-router restart.

If you are receiving non-v3 packets, then this is a peripheral bootup and values are assigned at the time.
there is no non-v3 requests in diag-router restart cases because peripherals are not capable of handling that.

So I believe handling diagid in this case should be careful without updating the global counter.

Reply:
**If v3 packets with diagid 0 is received, then this is a peripheral bootup and values assigned same as non-v3.
But if there is a value in v3, then its diag-router restart.** ---  Where is this logic handled? I don't find it in diag-router downstream code.
Just look at this part of the logic, the key point is whether this function will receive both V3 and non-V3 pkt. if there is only one type of pkt, then the logic here is fine.

Reply:
I would implement it to the spec that Manoj has laid out. I think downstream didn't handle it because we have control of which branch of diag-router runs with which firmware revision. That isnt quite the case with upstream. It's better to handle both versions per spec so that we can handle any combination of firmwares.

Reply:
I have made some modifications to this place according to your suggestions, please check it again in your spare time